krun/platform.py

```diff
@@ -191,9 +213,69 @@ def save_power(self):
                 warn("Could not set CPU%d governor to 'ondemand' when "
                      "finished.\nFailing command:\n%s" % (cpu_n, cmd))
 
+    def check_preliminaries(self):
+        """Checks the system is in a suitable state for benchmarking"""
+
+        self._check_taskset_installed()
+        self._check_cpu_isolated()
+        self._check_cpu_governor()
+        self._check_cpu_scaler()
+
+    def _check_taskset_installed(self):
+        from distutils.spawn import find_executable
+        if not find_executable("taskset"):
+            fatal("taskset not installed. Krun needs this to pin the benchmarks to an isolated CPU")
+
+    def _check_cpu_isolated(self):
+        """Attempts to detect an isolated CPU to run benchmarks on"""
+
+        # XXX Isolating the CPU will not prevent kernel threads running.
+        # We *may* be able to prevent atleast some kernel threads running,
+        # but it seems cset (the facility to do this on Linux) is currently
+        # broken on Debian (and perhaps elsewhere too). Bug filed at Debian:
+        # http://www.mail-archive.com/debian-bugs-dist%40lists.debian.org/msg1349750.html
+
+        with open(LinuxPlatform.KERNEL_ARGS_FILE) as fh:
+            all_args = fh.read().strip()
+
+        args = all_args.split(" ")
+        for arg in args:
+            if '=' not in arg:
+                continue
+
+            k, v = arg.split("=", 1)
+
+            if k != "isolcpus":
+                continue
+            else:
+                if "," in v:
+                    debug("Multiple isolated CPUs detected: %s" % v)
+                    vs = v.split(",")
+                    isol_cpu = int(random.choice(vs))
+                    debug("Chose (at random) to isolate CPU %d" % isol_cpu)
+                else:
+                    isol_cpu = int(v)
+                    debug("Detected sole isolated CPU %d" % isol_cpu)
+                break
+        else:
+            fatal("Krun failed to detect an isolated CPU!\n"
+                  "Did you add `isolcpus=X` to the kernel arguments?\n"
+                  "To do this on Debian:\n"
+                  "  * Edit /etc/default/grub\n"
+                  "  * Add the argument to GRUB_CMDLINE_LINUX_DEFAULT\n"
+                  "  * Run `sudo update-grub`\n"
+                  "When the system comes up, check `ps -Pef`.")
 
-    def check_cpus_throttled(self):
-        """Checks the CPU is configured for high performance
+        if isol_cpu == 0:
```

Comment:
This will only happen if the user actively chose to include CPU0 in the set of cores to isolate. The user chooses which cores by passing special arguments to the kernel at boot time via the bootloader, so it's really in the user's hands to make sure they don't isolate CPU 0.

[If the odds of choosing CPU 0 on a 4 CPU machine are 1/4, this means the user chose to isolate all four CPUs, thus not allowing anywhere for any non-benchmarking processes to run.]